omod/src/main/webapp/pages/results.gsp

```diff
@@ -1,75 +1,171 @@
+            sDom: 't<"fg-toolbar ui-toolbar ui-corner-bl ui-corner-br ui-helper-clearfix datatables-info-and-pg "ip>'
+        });
+
+        jq.fn.dataTableExt.ofnSearch['html'] = function (sData) {
+            return jq(sData).val();
+        };
+
+        jq('#rule-filter').change(function () {
+            resultTable.columns(0).search(jq(this).val()).draw();
+        });
+
+        jq("#patient-filter").autocomplete({
+            minLength: 0,
+            delay: 0,
+            source: Array.from(new Set(patients)),
+            close: function () {
+                if (jq(this).val() === "") {
+                    resultTable.columns(2).search("").draw();
+                }
+            },
+            select: function () {
+                resultTable.columns(2).search(jq(this).val()).draw();
+            }
+        });
+
+        jq('#free-text-search').keyup(function(){
+            resultTable.search(jq(this).val()).draw();
+        });
+
+        jq('#reset-filters').click(function(){
+            // Reset the table and clear all the filters
+            resultTable.search('').columns().search('').draw();
+            jq('#rule-filter').prop('selectedIndex', 0);
+            jq('#patient-filter').val("");
+            jq('#free-text-search').val("");
+        });
     });
     <% } %>
 </script>
-<% if (rule == null) {
-%>
-<h2>${ui.message("dataintegrity.result.all.label")}</h2>
+
+<% if (binding.variables.containsKey('rules')) { %>
+    <h2>${ui.message("dataintegrity.result.all.label")}</h2>
 <% } else { %>
-<h2>Results for ${rule.ruleName}</h2>
+    <h2>Results for ${rule.ruleName}</h2>
+<% } %>
+
+<% if (results.size() > 0) { %>
+    <div id="filter-area">
+        <% if (binding.variables.containsKey('rules') && rules) { %>
+        <label for="rule-filter">Rule:
+            <select id="rule-filter">
+                <option value="">Select rule</option>
+                <% rules.each { %>
+                <option value="rule_${it.ruleId}">${ui.encodeHtmlContent(it.ruleName)}</option>
+                <% } %>
+            </select>
+        </label>
+        <% } %>
+        <label id="patient-filter-label" for="patient-filter">Patient:
+            <input id="patient-filter">
+        </label>
+        <label id="free-text-search-label" for="free-text-search">Search:
+            <input id="free-text-search">
+        </label>
+        <button id="reset-filters">Clear</button>
+    </div>
 <% } %>
+
+<%
+def prettify = { "$it".replaceAll("null", "").replaceAll("\\s*,\\s", " ") }
+%>
 <% if (results.size() == 0) { %>
     There are no results to show. <a href="results.page?action=fireRules"><i class="icon-play"></i>&nbsp;Click here</a> to run all the validation rules, please note that it will take some time and slow down the server.
 <% } else { %>
 <table id="list-results" cellspacing="0" cellpadding="2" width="100%" class="dataTable">
     <thead>
     <tr>
+        <th style="display: none"></th>
```

Comment:
@ssmusoke , yes it is for rule filter and in the last sentence are you talking about patient filter ? I don't need patientId for filtering, datatable search from their names.